File: python/course-repo/solution/02-idempotency-and-crash-safety/src/webhooks/send_standalone.py
"""Fire a webhook delivery as a Standalone Activity and wait for it to finish."""

import asyncio
import sys
from datetime import timedelta

from temporalio.client import Client
from temporalio.common import RetryPolicy

from .activities import deliver_webhook
from .shared import WEBHOOK_RECEIVER_URL, TASK_QUEUE, WebhookDelivery


async def main(event_id: str) -> None:
    client = await Client.connect("localhost:7233")
    result = await client.execute_activity(
        deliver_webhook,
        args=[WebhookDelivery(
            url=WEBHOOK_RECEIVER_URL,
            payload={"event_id": event_id, "type": "order.created", "amount": 99.99},
            event_id=event_id,
        )],
        id=f"deliver-{event_id}",
        task_queue=TASK_QUEUE,
        start_to_close_timeout=timedelta(seconds=10),
        # Default RetryPolicy is unbounded — fine for the demo, dangerous in prod.
        # Cap attempts so a permanently-broken receiver doesn't retry forever.
        retry_policy=RetryPolicy(maximum_attempts=5),
    )
    print(f"Activity completed with status {result}")


if __name__ == "__main__":
    event_id = sys.argv[1] if len(sys.argv) > 1 else "evt_001"
    asyncio.run(main(event_id))

File: typescript/course-repo/exercise/02-idempotency-and-crash-safety/src/sendStandalone.ts
import { Client, Connection } from '@temporalio/client';
import { TASK_QUEUE, WEBHOOK_RECEIVER_URL } from './shared';

async function main(eventId: string) {
  const connection = await Connection.connect({ address: 'localhost:7233' });
  const client = new Client({ connection });

  const result = await client.activity.execute("deliverWebhook", {
    args: [{ url: WEBHOOK_RECEIVER_URL, payload: { eventId, type: 'order.created', amount: 99.99 }, eventId }],
    id: `deliver-${eventId}`,
    taskQueue: TASK_QUEUE,
    startToCloseTimeout: '10 seconds',
    retry: { maximumAttempts: 5 },
  });

  console.log(`Activity completed with status ${result}`);
  await connection.close();
}

const eventId = process.argv[2] ?? 'evt_001';
main(eventId).catch((err) => {
  console.error(err);
  process.exit(1);
});
